### ISSUE TITLE
Habitca Lag, Hanging, Glitches, Error Messages, Freezes

### ISSUE THREAD

**Issue by testsubject1987** [HabitRPG/habitica#6215]
I am using Habitica with Firefox (42.2) on a Windows XP3 system and also on a Windows 8.1 system, the problem is more pronounced with greater frequency on the XP system, when scrolling or clicking around the whole site will lag, hang, glitch, give error messages, sometimes freezes the browser ranging from 2-6 seconds on the 8.1 system and 5-20 seconds each tick on the XP system.

The problem I have noticed on the XP system for... nine months or more... this past month I've also noticed it on the 8.1 system so I am here... It will happen for all items from clicking links, items, to-dos, habits, dailies, rewards... 149 Habits, 211 ToDos, 4 Rewards

Both my XP and 8.1 system are very comparable in specs and both otherwise run fine. The XP system has less graphics memory. However... 6-9 months ago this problem I really didn't notice and the site ran very smoothly almost all the time and the amount of items had was perhaps a little bit less but not so much less (I mean if it is then my bad, but I dont think its increased that much).

I think its important to note that this XP system is otherwise running fine and most sites open as quickly and smoothly as on the 8.1 system... and its really only habitica that does this for me.

**Comment by gauthamchandra** [comment]
@testsubject1987 Can you give us what the specific error messages are?

If you need help finding browser console errors (errors not shown to your directly by Habitica), Go to Firefox's Tools -> Web Developer -> Toggle Tools. It should toggle a developer toolbar that looks something like this:

<img width="1434" alt="screen shot 2015-11-11 at 2 52 37 pm" src="https://cloud.githubusercontent.com/assets/5430280/11101258/fca871c8-8883-11e5-8d73-be087321013d.png">

Can you tell us what errors might be there?

**Comment by testsubject1987** [comment]
Hello,

I apologize for the delay. Its not exactly something I like wrangling with. I get frequently a script error at 'Script: https://habitica.com/app-c96340b6.js:5'

Here is an attached screenshot of what you asked for.

![consolehabitica20151115pm0229](https://cloud.githubusercontent.com/assets/15749182/11170821/61db6978-8ba5-11e5-9e3c-b985e1492c3f.jpg)

I hope that helps further any advise and resolution?

**Comment by gauthamchandra** [comment]
Not 100% sure yet but this seems to be purely a performance issue.

> 149 Habits, 211 ToDos, 4 Rewards

If I had to guess, I would say this is due to a lot of the dailies, tasks and todos you have causing the `$digest` cycle to hoard the JS event loop and forcing Firefox to just kill the script.

I am going to do some testing to see if I can boost performance.

**Comment by testsubject1987** [comment]
It would be my guess to IF the two computers I am using were not so similar in specs or if the problem was constant over a long period of time.

To note, perhaps I should of said this, that consoling was done before I clicked or did anything. Just loaded it up and clicked into it. Not sure if that makes any difference.

I have a few hunches... my 8.1 and XP systems have a difference... and I think it may have to do with my graphics memory on the XP system (though that doesn't explain why its slowly become a problem over my use of habitica, I've been using Habitica for more than two years now?) OR bigger hunch that its Windows XP all together... When I update programs I get messages to the effect that it may not work correctly because Microsoft took XP, put it in a bag, and threw it into a river... This system would need some hardware updates to move to XP and candidly I just like XP... I've had it for so long I am a bit stubborn in my ways... and as such I am a holdout. Another hunch is maybe (though I turn adblocker off and tried it) is that it may have to do with the adblocker program I use. Though I can't imagine how it would be causing these errors because the 8.1 system has it too and its filters are almost identical?

If it really is just the amount in my habitica I can change that... or at least I can plan accordingly. But it just strikes me as a little bit odd that if that is the case that the 8.1 system does okayish with it? I'm more or less trained to hit keys on a computer, I'm not so great with how they actually operate, so I yield to better expertise.

Thank you for your efforts and I hope it helps... otherwise I may have to buckle down and get something like a tablet just to use habitica... which is a bit expensive for my budget right now. It just seems like it should be able to work on XP? Maybe?

**Comment by gauthamchandra** [comment]
@testsubject1987 

> I think it may have to do with my graphics memory on the XP system (though that doesn't explain why its slowly become a problem over my use of habitica, I've been using Habitica for more than two years now?)

Browsers use your GPU to do some hardware acceleration for rendering for added performance. The lack of graphics hardware on XP _could_ add to the problem but I personally think its mostly the number of dailies, todos and habits you have. The framework that Habitica uses for JS (Angular 1.x) doesn't work well with large amounts of data unless it is tuned properly. **I am not too familiar with the codebase so I will have to spend this weekend to look into this issue so I can't confirm my theory just yet**

> Another hunch is maybe (though I turn adblocker off and tried it) is that it may have to do with the adblocker program I use.

Yup. Please try testing with all script blocking plugins like NoScript, Disconnect, Ghostery, uBlock, Adblock plugins disabled or whitelisted.

> When I update programs I get messages to the effect that it may not work correctly because Microsoft took XP, put it in a bag, and threw it into a river

End of life for Microsoft Support for XP afaik was April 2014.

**tl;dr**: I am going to look more into boosting performance but no promises :+1:

**Comment by testsubject1987** [comment]
How goes it?

**Comment by Alys** [comment]
@testsubject1987 We can't really support XP since Microsoft has ended its support for it. gauthamchandra's advice to test with all script blocking plugins like NoScript, Disconnect, Ghostery, uBlock, Adblock plugins disabled or whitelisted is good - did you try that?

**Comment by testsubject1987** [comment]
It happens on the non-xp system as well. I have no script blocking plugins, disconnect, ghostery, or ublock. I tried disabling and uninstalling adblock and the same problem persists.

**Comment by Alys** [comment]
@testsubject1987 Please create a second account with some test tasks in it (just a few) and use it on the same browser that you use your real account on. Is it noticeably faster? If so, the problem is probably the large number of tasks you have, as suggested by gauthamchandra. Unfortunately there's not a lot we can do about that right now, but we are in the middle of a large project to make some significant site-wide  improvements to the back-end code and database structure. That's likely to produce some performance improvements in some areas of the site, and will open up possibilities for more improvements in future.

**Comment by gauthamchandra** [comment]
Yeah. I got super busy at work and was unable to work on this recently. :(

I can look into some short term fixes this coming weekend.

**Comment by Alys** [comment]
@testsubject1987 I'm closing this since it's been a while since there was an update from you, but if you found anything interesting from the test I suggested in my previous comment, please tell us.
